### DISCUSSION TITLE
Memory usage with ReadSample

### DISCUSSION THREAD

**Discussion by Ciaran3** [FactoryXCode/MfPack#14] [Q&A] [answered]
Hi all,
Thanks for the great work on keeping this library maintained! Sorry I have not contributed in a long time..
I have a question regarding using ReadSample on the source reader - with asynchronous mode. 
Essentially: (https://docs.microsoft.com/en-us/windows/win32/medfound/using-the-source-reader-in-asynchronous-mode)
I'm just capturing video frames, so I am not selecting the audio stream.

I've read your discussion on the topic:
https://github.com/microsoft/Windows-classic-samples/issues/108
https://social.msdn.microsoft.com/Forums/en-US/c2b39d7b-5203-492c-9663-75d1601f82c7/memory-leak-imfsourcereaderreadsample-method?forum=mediafoundationdevelopment

I'm using critical section and releasing the sample (=nil), but still the memory usage seems to keep going up.. Even after the source reader is flushed and closed the memory usage does not seem to free up completely.

Is this still a known memory issue with media foundation? Or perhaps I'm still missing something. 
If the source reader is caching the samples, I would expect these to be released at least on Flush.

Admittedly i'm using version: 2.6.1, so upgrading to 3.0.1 will be my next step, in case some calls have been updated to fix this.

**Comment by FactoryXCode** [comment, marked as answer]
Hi,
To fix this issue, I implemented MsgWaitForMultipleObjects to give the
ReadSample method more time to organize it's sample pool with in the OnReadSample  function, alternatively you can use the Sleep() method.
 
See: Example MfAudioClipEx (function TAudioClipClass.OnReadSample)

Although I think you need to implement a callback function from IMFMediaEvent, that reports when a sample is processed, before processing the next one, as usual in a-synchronous mode. I think function IMFMediaEvent.GetStatus(out phrStatus: HRESULT) could do the job.

Right now I'm working on a WASAPI component that should render RAW/WAV formats. (See: Examples\MFComponents\WasApiRenderFile.pas. That is not working yet.

Greeting, Tony.

**Comment by Ciaran3** [comment]
Hi Tony,

Thanks for your reply. Unfortunately, I haven't had any luck with that solution either.
I've even tried testing with the synchronous ReadSample instead of asynchronous, to narrow down the issue - but still no luck completely freeing up the memory.

This example below is the memory in use, when reading a single video frame sample.

- Memory usage before calling ReadSample: 10 MB
- After ReadSample: 39 MB (Makes sense, the sample is in memory)
- After releasing the sample and flushing the source reader: 40 MB (What??)
- After destroying the IMFSourceReader: 16 MB

As you can see, the memory is **nearly** freed up after destroying the source reader, but not before.. 
A real shame if the source reader needs to be closed to free up the memory. A use case of requesting multiple video thumbnails, at different positions in the video will be much slower if the source reader needs to be closed and reopened each time. 
Unless Media Foundation does it own memory clean up at a certain point? But I cant find any details on that, nor anywhere where I can specify the cache size.

Edit - I can put together a sample project demonstrating this, if that would be of use.

**Reply by FactoryXCode** [reply]
I agree it's a weird issue. I asked MS about the cpu usage, but that is another thing. However, a debugger like MadExcept does not report any issues about memory leaks, nor does mstrace. Did you close com and MF ? 

`function CloseMF(): HResult;
begin
  // Shutdown MF
  Result := MFShutdown();
  // Shutdown COM
  CoUninitialize();
end;`

**Reply by Ciaran3** [reply]
Yeah I guess its not exactly a leak, as the memory is freed eventually, just not when I want (flushing the source) 😞 
I'll rework my asynchronous code into the demo project and see if that resolves the issue. Perhaps its an issue with my current implementation 🤞

**Comment by Ciaran3** [comment]
Hi Tony,

I've added a sample frame capture project here:
https://github.com/Ciaran3/MfPack/tree/Master/MfPack/Samples/MfFrameCapture

Let me know if you see any ways to improve it. Ideally, I'll update it to have an asynchronous capture example as the memory issue does not seem as bad with this synchronous code.

If you think this sample project is of use for MfPack, I can open a pull request 👍 

![MfFrameCaptureSample](https://user-images.githubusercontent.com/31159631/133241055-744dc051-b3a0-4ec0-92ce-fcd302232143.png)

**Reply by FactoryXCode** [reply]
That would be great. Thank you. At the moment every one is busy, I'm working on a new sample and have to study the new windows 11 sdk. Thanks again for your efforts. Just do that pull request.

**Reply by Ciaran3** [reply]
No problem, good to give back a bit to this project 👍 I'll see if I can get it working asynchronously as well.

**Reply by FactoryXCode** [reply]
Ok, I did take a short tour. First: A TBitmap should be released from memory with FreeAndNil(your bitmap). Second: The size of the samples are a factor to 10 high. 50 mb should be 5 mb and so on. Your sample does what it supposed to do. A frame from a compressed frame can vary from 4 up tp 9 mb, depended of colors etc. You can check those variations when you take a picture on HD on your smartphone. When you save a snapshot to file, you will see the difference.  Further on: The interface will create a sample pool of 10 samples. Maybe this is why the memory token is 10 times higher. I will update your sample soon, with adjusted code. Greetings Tony.

**Reply by Ciaran3** [reply]
Thanks for that! Looking forward to seeing your suggested changes. I'm just refactoring and getting the asynchronous code working, but I will also incorporate your changes.

**Reply by FactoryXCode** [reply]
Ok, I think I fixed the issue. Memory usage is more consistent now. Changes are committed to Master right now.

**Reply by Ciaran3** [reply]
Excellent, thanks. I'll have a proper look through your changes on the weekend. So it looks like SafeRelease is needed instead of just =nil? 
Now I'm just looking at the asynchronous code and trying to figure out the safest method to return the image from the worker thread (OnReadSample) to the main thread. Otherwise, return a reference to the sample and let the main thread do the image processing..

**Reply by FactoryXCode** [reply]
I always release COM interface pointers with this procedure. Delphi internally treat them as pointers. An interface declared in a method does not have to be released, because they will be released automatically (C++ compilers don't). Also you don't have to Nil them before using the interface (In C++ you have to do that) Global interfaces, however, should be released by SafeRelease(). For an example of getting the image or frame from OnReadSample, you could see in function TAudioClipClass.OnReadSample in example MfAudioClip(Ex). I think it doesn't matter if this is an audio or video frame.

**Reply by FactoryXCode** [reply]
Maybe a handy feature is to take snapshots from a hardware(video) or file?

**Reply by Ciaran3** [reply]
> Maybe a handy feature is to take snapshots from a hardware(video) or file?

That sounds like a good idea. I haven't looked into capturing from a device, but hopefully a good amount of the existing code code can be reused.

**Reply by Ciaran3** [reply]
> I always release COM interface pointers with this procedure. Delphi internally treat them as pointers. An interface declared in a method does not have to be released, because they will be released automatically (C++ compilers don't). Also you don't have to Nil them before using the interface (In C++ you have to do that) Global interfaces, however, should be released by SafeRelease(). For an example of getting the image or frame from OnReadSample, you could see in function TAudioClipClass.OnReadSample in example MfAudioClip(Ex). I think it doesn't matter if this is an audio or video frame.

Ah yes, that makes sense as the way to release global COM interface pointers. I'll check out that example 👍🏼 I'm just aware cross thread access to TBitmap has been an issue in the past.

**Reply by FactoryXCode** [reply]
OK, I studied your sample.  That being said, you should separate the users interface (form)  apart from the interfaces. So, there should be a class unit for sync and async mode and a unit that holds shared methods they both use.  In that way it's easier to read and maintain about what's going on. The big question is, how to transport the bitmap. For this send a message to the form and copy the public bitmap from or the sync or async class. Remember using async handling, you have to set commands indirectly by setting up the commands (requests) on the appropriate handlers. See sample MfpPlayerClass.pas of mfplayer 1. There is no way to send direct calls like you can do in sync without code going to be crash. An other thing; use HResult return values to evaluate a function. This will things make easier to debug.  You did a great job, thank you!

**Reply by Ciaran3** [reply]
Thanks for that 👍 I agree that it is getting messy with the mix of async and sync code as well as the source reader setup on the form. I'll refactor, do a pull request and then work on the other things :)

**Reply by FactoryXCode** [reply]
Thanks for that. i did the request and noticed the Async is not finished yet. I think if you reinitialize the FRenderTarget and adding the same code from procedure TFileCaptureSync.ProcessSample and TFileCaptureAsync.ProcessSample it should work? added some suggestions to the source.

**Reply by Ciaran3** [reply]
Thanks, I'll have a look at your latest suggestions.
Oh the async capture should be working in the latest code, because the sample converter is setup in the create of TFileCaptureAsync. Let me know if you still are having issues, it might be an issue between our builds.

**Reply by FactoryXCode** [reply]
Ok, thanks. I was a pull behind. Should I put the mfpack headers on top of the files? 
I've some suggestions, if you don't mind. 
- Maybe it's more practical to display the log within the IDE (debugwindow)?
- A save bitmap to file option? (see: https://docwiki.embarcadero.com/RADStudio/Sydney/en/Saving_a_Picture_to_a_File)
I will test more for memory usage.

**Reply by Ciaran3** [reply]
No problem, glad I hadn't missed something.
- Logging - Yes, that would probably be better in the long run. For now when testing, I just like seeing the logging in the same spot to make it easier to track capture times and memory usage 👍 
- Adding a button to save to captured bitmap is definitely a good idea. I can add that.

**Reply by FactoryXCode** [reply]
For some reason a memory leak of 16 bytes persists in procedure TFileCaptureAsync.ProcessSample.
When closing the application oSampleReply stays in memory it seems..

**Reply by FactoryXCode** [reply]
Ok, I did solved the memory leaks in Asynchronous.pas. I had to change PostMessage to SendMessage.

**Reply by FactoryXCode** [reply]
However, I'm not sure if that's the right way.  Maybe something in the message queue is not going the way it supposed to be in the a-sync handling.

**Reply by Ciaran3** [reply]
Thanks for looking into this, I will have a look at your changes. Does it happen consistently for you, or it a timing issue, between the frame request and then closing the application? FastMM is not reporting anything for me, but its not super reliable..
I can see a potential issue - I'm not checking if FMessageHandler.Handle is valid before posting any of the messages. This should probably be checked before any message is created - in the case of the application closing just as a frame has just been found.
PS - Do you use a code formatting profile? Maybe we should use the same one to make merging/comparing changes easier 👍

**Reply by FactoryXCode** [reply]
The leak happens when using A-sync after every "shot".  
I added a screenshot, where you can see after taking 2 "shots", 2 leaks in separate threads etc.
![image](https://user-images.githubusercontent.com/41634066/134633255-9999bf34-5819-4c34-a1e2-9e53c220423c.png)

I use madExcept, which is quite good. MfTrace I use so now and then.

I do not use a formatter. When we started this project (11 years ago) we agreed to use the same format as used in the headers as much as possible to make reading and debugging code  easier. Also in the past we agreed with Henry Gourvest (DSPack) to make it as difficult as we can to prevent Embarcadero kidnap the project like happened with DSPack.

**Reply by FactoryXCode** [reply]
Ok, The issue is -I think- the pointer to oSampleReply in method TFileCaptureAsync.ProcessSample will never be released if declared in this method. Now, moving the declaration as a private class var and declare it as oSampleReply: TSampleReply; will do the job it seems..
So there is also no need to dispose, because the object wil be released automatically from memory by the compiler.
While I'm  still worried about the amount of used memory.

**Reply by Ciaran3** [reply]
I've just given the latest version a go (from the Sandbox folder), looking really good Re - Memory usage. Nice work 👍 Interesting solution regarding the TSampleReply. I guess that makes sense, and simplifies the handling of the memory, as we should only ever need a single reference to TSampleReply. 

> I use madExcept, which is quite good. MfTrace I use so now and then.

Yeah that looks much more detailed - definitely will look into using that.

> I do not use a formatter. When we started this project (11 years ago) we agreed to use the same format as used in the headers as much as possible to make reading and debugging code easier. Also in the past we agreed with Henry Gourvest (DSPack) to make it as difficult as we can to prevent Embarcadero kidnap the project like happened with DSPack.

Haha, fair enough. I understand the precaution.

**Reply by Ciaran3** [reply]
I'm seeing an error when capturing from a WMV with the latest code (both sync and async). How about you? 

I grabbed a sample video from here: https://filesamples.com/formats/wmv
Then after a few frame grabs, an AV occurs: 
![call-stack](https://user-images.githubusercontent.com/31159631/135038288-af90b9f4-074d-4fb0-a871-80522f9dd3d5.PNG)

**Reply by FactoryXCode** [reply]
I noticed. I tested on a mp4 and everything is going ok. Yet i'm rendering to a wmv to see if the files on that site are not corrupted.

**Reply by FactoryXCode** [reply]
Ok, I tried to play those files from  https://filesamples.com on VLC and the Windows Media Player ( they render like shit -sorry for the language- anyway) a more reliable sample also generates the same pointer error in mfplat.dll, after taking one sample. So what I asked on MSDN, there is something wrong with this method internally, I think. Maybe there is a wrong order of methods in the interface (I had that before, see issues) I will investigate that. Any clue from your side?

**Reply by Ciaran3** [reply]
I think it is partly to do with the confusion between WinApi.DirectX.D2D1 and WinApi.D2D1. My local code is working with those videos, and that is one of the differences i can see . I'm using the WinApi.D2D1 (with Delphi 11). Will let you know if I figure it out.

**Reply by Ciaran3** [reply]
Ah no - its not the D2D1 units (although that does cause confusion).
Its actually **RemoveAllBuffers** that is the cause - remove that an no error is raised. I remember running into that issue last time I tried that call, so I just dont use it. Even without it memory usage still seems pretty good in my local copy.

**Reply by FactoryXCode** [reply]
That is good news 👍 
Still weird this exception only occurs while reading samples from .wmf files.

**Reply by FactoryXCode** [reply]
_Ah no - its not the D2D1 units (although that does cause confusion)_

The MfPack D2D1 API has a different  unit scope (WinApi.DirectX.D2D1) to prevent name mangling with the APi included with Delphi (WinApi.D2D1) . Since 2020 D2D1 had major updates that Embarcadero did not implement. There is no way  those versions can be mixed unless both are declared, which is not a good idea. Also, older Delphi versions (<2019) don't have this API available. Remember when writing code, it must be possible to compile from Delphi XE2 and up.

**Reply by Ciaran3** [reply]
> Also, older Delphi versions (<2019) don't have this API available. Remember when writing code, it must be possible to compile from Delphi XE2 and up.

ah yes, that does make it tricky. Do you have multiple Delphi versions to check compatibility? Or you just know by now the areas to check? In my own copy I'm just using the WinApi.D2D1 in the SampleConverter unit - but I guess that might cause issues elsewhere if it is missing major updates.

**Reply by FactoryXCode** [reply]
> Do you have multiple Delphi versions to check compatibility?

Yes I do. But usual I check with XE7, that version, in my opinion, is also safe to check for older versions.
Further, you can see at the date in the units, when an API is introduced. For instance the Embarcadero API's D2D1 and DXGI are not working with some samples, like MfThumbnails. Embarcadero has a bad reputation to maintain the API's they release or there is no API at all (like Media Foundation now and DirectShow in the past) or the API is incomplete (WinMM API).

> but I guess that might cause issues elsewhere if it is missing major updates.

All API's included with MfPack are up to date with the latest SDK. When Microsoft is releasing a new sample, it almost always need the latest SDK So there's a risk Delphi is missing parts or a whole API.

In the meanwhile I added png and jpg functionality for saving to file. These API's where introduced in Delphi 2009.

I checked the MFSample interface, but according to the vtables in MFPlat.dll, there is nothing wrong in sequence or declaration. So, the RemoveBuffers issue only affects VFW right? Because I did test with AVI and MP4, and there it will not raise any pointer exception.
